wallet: "avoid_reuse" wallet flag for improved privacy

kallewoof opened the pull request:
Add a new wallet flag called `avoid_reuse` which, when enabled, will keep track of when a specific destination has been spent from, and will actively "blacklist" any new UTXOs which send to an already-spent-from destination.

This improves privacy, as a payer could otherwise begin tracking a payee's wallet by regularly peppering a known UTXO with dust outputs, which would then be scooped up and used in payments by the payee, allowing the payer to map out (1) the inputs owned by the payee and (2) the destinations to which the payee is making payments.

This replaces #10386 and together with the (now merged) #12257 it addresses #10065 in full. The concerns raised in https://github.com/bitcoin/bitcoin/pull/10386#issuecomment-302361381 are also addressed due to #12257.

~~Note: this builds on top of #15780.~~ (merged)

kallewoof committed:
wallet: avoid reuse flags

Add m_avoid_address_reuse flag to coin control object.
Add avoid_reuse wallet flag and accompanying strings/caveats.

promag commented on test/functional/feature_avoidreuse.py:
```diff
@@ -0,0 +1,92 @@
+        '''
+        balance = node.getbalance()
+        if balance > 0.5:
+            tosend = '%.5f' % (balance - Decimal(0.01))
```
Commit "test: Add test for avoidreuse feature"

Use `subtractfeefromamount=true` below instead?

promag commented on src/wallet/init.cpp:
```diff
@@ -55,6 +55,7 @@ void WalletInit::AddWalletOptions() const
 {
     gArgs.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), false, OptionsCategory::WALLET);
     gArgs.AddArg("-avoidpartialspends", strprintf(_("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it after spending from it), but may result in slightly higher fees as suboptimal coin selection may result due to the added limitation (default: %u)"), DEFAULT_AVOIDPARTIALSPEN ...
+    gArgs.AddArg("-avoidreuse", "Mark addresses which have been used to fund transactions in the past, and avoid reusing these in future funding, except when explicitly requested " + strprintf(_("(default: %u)"), DEFAULT_AVOIDREUSE), false, OptionsCategory::WALLET);
```
IMO could be something like `-avoidaddressreuse`.

promag commented on src/wallet/coincontrol.h:
```diff
@@ -34,6 +34,8 @@ class CCoinControl
     //! Avoid partial use of funds sent to a given address
     bool m_avoid_partial_spends;
+    //! Allows inclusion of dirty (previously used) addresses
+    bool m_allow_dirty_addresses;
```
Commit "wallet: add m_allow_dirty_addresses flag to coin control object"

Could initialize here?

kallewoof replied:
Everything is initialized in `SetNull()`. In some cases, you can't initialize in the `.h` file because no `gArgs` available at compile time.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -922,6 +922,37 @@ bool CWallet::MarkReplaced(const uint256& originalHash, const uint256& newHash)
+    if (srctx) {
+        CTxDestination dst;
+        if (ExtractDestination(srctx->tx->vout[n].scriptPubKey, dst)) {
+            AssertLockHeld(cs_wallet);
```
Commit "wallet: enable avoidreuse feature"

Why is this here? Move up?

Same in `CWallet::IsDirty` below.

kallewoof replied:
This may be artifact from previous iteration. Removing.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -922,6 +922,37 @@ bool CWallet::MarkReplaced(const uint256& originalHash, const uint256& newHash)
+void CWallet::SetDirtyState(const uint256& hash, unsigned int n, bool dirty)
+{
+    const CWalletTx* srctx = GetWalletTx(hash);
+    if (srctx) {
```
Commit "wallet: enable avoidreuse feature"

Instead of nesting, could early return here:

```cpp
if (!srctx) return;
```

Same in `CWallet::IsDirty` below.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1940,6 +1979,7 @@ CAmount CWalletTx::GetImmatureCredit(bool fUseCache) const
 
 CAmount CWalletTx::GetAvailableCredit(bool fUseCache, const isminefilter& filter) const
 {
+    bool allow_dirty_addresses = !gArgs.GetBoolArg("-avoidreuse", DEFAULT_AVOIDREUSE);
```
Commit "wallet: enable avoidreuse feature"

Could add a `g_allow_dirty_addresses` instead?

kallewoof replied:
Might be worth it, yeah.

promag reviewed (commented):
Concept ACK.

Should we also disallow sending to a dirty address? If we don't then those coins can't be spend if the flag is set, or am I missing something?

practicalswift commented:
Concept ACK

Nice work!

gmaxwell commented:
Perhaps fodder for a separate PR,  but listtransactions and the GUI should get indicators that an incoming payment was dirty when received.  By having that parties that reuse addresses will start getting a frowny-recycle-icon or whatever on their payments, which will increase awareness that their behaviour has downsides.

kallewoof commented:
@promag 

> Should we also disallow sending to a dirty address? If we don't then those coins can't be spend if the flag is set, or am I missing something?

You can always spend them by using the `allowdirty` flag or by manually selecting them using coin control (via `createrawtx` or via the GUI).

I do think we should at least warn users about that kind of behaviour though, but it feels like a UI fix that can come later.

@gmaxwell 

> Perhaps fodder for a separate PR, but listtransactions and the GUI should get indicators that an incoming payment was dirty when received. By having that parties that reuse addresses will start getting a frowny-recycle-icon or whatever on their payments, which will increase awareness that their behaviour has downsides.

I will see about making that the case for the CLI side of things. I think a GUI side fix would be great, but unfortunately I don't seem to be talented at writing QT code. (Practice, I guess.)

> Edit: I see I'm repeating myself from an earlier PR. :P

Sorry about that. I felt like it was worthwhile to re-PR since the old PR has a lot of outdated talk that is mitigated by #12257.

Sjors commented:
@gmaxwell:
> the GUI should get indicators that an incoming payment was dirty when received

I think "dirty" is a confusing concept. Maybe add an exclamation mark (or a detective icon) next to the transaction and when the user clicks on that, say something like "This address has been used before, for privacy reasons it's better to create a new address each time you wish to receive coins, even from the same person".

Detective icon is a nice hint that there's privacy related information, but an exclamation mark can also be used to explain other potential problems with an incoming transaction, e.g. if the fee is extremely low.

A parallel measure, that doesn't involve UI changes, could be for coin selection to not spend from dirty addresses as long as possible. And then try to spend them if there's an exact match (no other outputs, no change), perhaps even only if there's no other exact match, depending on how strongly you want to avoid spending these things.

When the user wants to "spend all" funds and the dirty amount is less than x%, a dialog could popup and suggest to "exclude a small amount for privacy reasons".

This is the kind of behavior that if we want people to use it, should be on by default. That can wait for another PR, but it's useful to think about it a bit when designing the RPC.  `-avoidreuse` doesn't really capture the above. It's too aggressive / binary an option to be useful in a GUI IMO.

Telling people to manually go into coin selection is not a good idea. If a user receives a non-trivial amount, they expect to be able to just spend it with no additional hoops. That means in the current form it's probably never a user friendly default.

promag commented:
> Maybe add an exclamation mark

@Sjors I like that, and then a tooltip/popup could explain the warning.

> `-avoidreuse` doesn't really capture the above. It's too aggressive / binary an option to be useful in a GUI IMO.

"Avoid" is not binary right?

kallewoof commented:
> I think "dirty" is a confusing concept.

It's called "dirty" in the CLI as well, but there's also a different concept in the wallet code called 'dirty' which is completely unrelated. In short, it may be a good idea to rename this feature, but I can't think of a good name. "Compromised" is too long. "Seen" is too vague. @luke-jr any ideas?

> Detective icon is a nice hint that there's privacy related information, but an exclamation mark can also be used to explain other potential problems with an incoming transaction, e.g. if the fee is extremely low.

I like "!" too. If it is used for multiple things, it could take away from the importance though (e.g. user ignores it thinking it's 'just cause of the 1 sat/b fee').

As a sidenote, we also should mark UTXOs in coin selection dialog somehow. Same approach? Maybe need to include a warning popup if they mix dirty and clean.

> A parallel measure, that doesn't involve UI changes, could be for coin selection to not spend from dirty addresses as long as possible. And then try to spend them if there's an exact match (no other outputs, no change), perhaps even only if there's no other exact match, depending on how strongly you want to avoid spending these things.

I don't know, I think we should always avoid dirty inputs unless the user explicitly marks them as 'clean' or picks them directly using manual coin control.

> When the user wants to "spend all" funds and the dirty amount is less than x%, a dialog could popup and suggest to "exclude a small amount for privacy reasons".

I didn't think about this case, and it's probably fairly common. Then again, I don't think I can do a lot from the CLI side, so this is probably GUI level stuff.

> This is the kind of behavior that if we want people to use it, should be on by default. That can wait for another PR, but it's useful to think about it a bit when designing the RPC. -avoidreuse doesn't really capture the above. It's too aggressive / binary an option to be useful in a GUI IMO.

What about it is too aggressive?

> Telling people to manually go into coin selection is not a good idea. If a user receives a non-trivial amount, they expect to be able to just spend it with no additional hoops. That means in the current form it's probably never a user friendly default.

Maybe it should be sensitive to the amount. A good middle ground may be to, for any input that is X times higher than current dust threshold, that goes to an already spent-from address, to show the user a dialog saying the input is dirty and ask them whether to force-mark it as clean or keep it as dirty.

Sjors commented:
> What about it is too aggressive?

A user might receive their entire salary on a reused address. If the current implementation of `-avoidreuse=1` were to become the default, then the UI would need to honor that setting. That means at minimum asking the user for confirmation when they're about to spend these funds. But that would be a very unintuitive question. The Spend screen is not the right place to handle this. Warning the user when they receive such funds is more appropriate.

This is where the threshold comes in, but I think it's non-trivial to figure out what the right threshold is, and it might be game-able.

Perhaps the ! / detective icon in the transaction view could offer the user a choice to manually quarantine funds. Something like "If you did not expect this transaction someone may be spying on you, waiting for you to spend the coins. Would you like to quarantine them?"

So then there's `dirty` with a subset of `quarantined`. `-avoidreuse=1` could e.g. not spend `quarantined` without an override, give other `dirty` coins special treatment, but still spend them if needed. That way the setting can be made a default in a future version, in way that the GUI can honor it.

Depending on what ends up happening, the documentation should perhaps point out that the GUI ignores this setting. Alternatively (for the current implementation) dirty coins could be hidden in the GUI, not shown in coin selection and not used when spending all, when this setting is enabled. Both options require a warning, and neither seems ideal, but updating the GUI in a more sophisticated way is pretty time consuming.

kallewoof commented:
@Sjors That makes sense to me. It seems like adding this default off (as is the current proposal) makes the most sense.

GUI work seems like it will be a good deal of work to get right, but in the meantime, there are real (advanced) users who would benefit from having this feature now, even without the GUI/intuitive component.

luke-jr commented:
N.B. This doesn't actually fully solve #10065, since transactions received with a dirty address are still shown in the GUI / RPC.

luke-jr commented:
(as for a term... "reused" perhaps? I don't know that this flag should be exposed as-is, but more like a "will never confirm until spent" status)

kallewoof commented:
I think "reused" is a bit vague, and doesn't convey the fact it's considered a thing to be avoided if possible, in the way that "dirty" does.

kallewoof commented:
Note: I realized the test was invalid, as `getbalance` would return only clean amount, so I added support to `getbalance` and fixed the tests.

kallewoof commented:
There is an alternative #13801 to this PR, which replaces the `dirty=true/false` flag with a `dest_filter=mixed/clean/dirty` filter. I will close one of these based on feedback on which one people prefer.

jnewbery commented:
It feels to me that this should be a persistent per-wallet setting, rather than a global config option (I think we should be eliminating the global wallet config options wherever possible. See #13044)

jnewbery commented:
> a PR @jonasschnelli was working on that wasn't merged yet (the no private keys one)

#9662 - merged!

kallewoof commented:
Cool! I was trying to find that one.

kallewoof commented:
After @gmaxwell comment [here](https://github.com/bitcoin/bitcoin/pull/13801#issuecomment-409262305), the new funding approach would work as follows:

1. Create output groups of all dirty coins
2. See if any of them fulfil the desired payment amount including fees up to maybe 3x the amount, and return that group if found. MARK CHANGE AS DIRTY
3. Try coin select using only clean coins

If user used `-allowdirty` flag, it would simplify to

1. Try coin select using all coins. (MARK CHANGE AS DIRTY IF 1+ DIRTY COINS WERE USED)

Currently the code does step 3 of the first case and partially step 1 of the `-allowdirty` case, but does not mark change as dirty. So:

* [ ] Switch to using persistent per-wallet mode
* [ ] Add check to see if dirty output groups would satisfy payment and use that if so
* [ ] Mark change outputs with 1+ dirty coins as their origin as also dirty

kallewoof committed:
wallet: make IsWalletFlagSet() const

kallewoof commented:
@jnewbery Code is now using the wallet flag feature. The `-avoidreuse` flag now determines the default value for new wallets instead. (Also ping @jonasschnelli regarding wallet flag stuff.)

practicalswift commented on test/functional/feature_avoidreuse.py:
```diff
@@ -0,0 +1,129 @@
+        '''
+        Test the simple case where [1] generates a new address A, then
+        [0] sends 10 BTC to A.
+        [1] spends 5 BTC from A. (leaving roughly 4 BTC useable)
```
Typo found by `codespell`: useable

kallewoof replied:
> Usable can also be spelled useable, with an e in the middle: both are correct.

https://english.stackexchange.com/questions/69709/is-useable-preferred-in-certain-regions-or-just-an-alternate-spelling-of-usa

jnewbery commented on src/wallet/init.cpp:
```diff
@@ -55,6 +55,7 @@ void WalletInit::AddWalletOptions() const
 {
     gArgs.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), false, OptionsCategory::WALLET);
     gArgs.AddArg("-avoidpartialspends", strprintf(_("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it after spending from it), but may result in slightly higher fees as suboptimal coin selection may result due to the added limitation (default: %u)"), DEFAULT_AVOIDPARTIALSPEN ...
+    gArgs.AddArg("-avoidreuse", "Enable the 'avoid reuse' flag for new wallets created. When enabled, wallets will mark addresses which have been used to fund transactions in the past, and avoid reusing these in future funding, except when explicitly requested " + strprintf(_("(default: %u)"), DEFAULT_AVOIDREUSE), false, OptionsCategory::WALLET);
```
My personal preference would be to remove this command line argument, since it's possible to create a wallet with this flag using the `createwallet` RPC, or update an existing wallet's flag with `setwalletflag`. I think it's better to remove it from here because:
- it's better to have one way to do things
- in general, I think we should be aiming to reduce the number of command line arguments, especially in the wallet (see https://github.com/bitcoin/bitcoin/issues/13044)

kallewoof replied:
Makes sense. And resolves one TODO (the startup option is ignored for "fresh" nodes in the test framework).

jnewbery commented on src/wallet/init.cpp:
```diff
@@ -227,8 +228,11 @@ bool WalletInit::Open() const
         return true;
     }
 
+    uint64_t wallet_creation_flags =
```
I'd prefer to remove this, and make the flag settable through the `createwallet` and `setwalletflag` options (see how `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set)

kallewoof replied:
Above change resulted in its removal.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -292,7 +292,7 @@ static UniValue setlabel(const JSONRPCRequest& request)
 static CTransactionRef SendMoney(CWallet * const pwallet, const CTxDestination &address, CAmount nValue, bool fSubtractFeeFromAmount, const CCoinControl& coin_control, mapValue_t mapValue)
 {
-    CAmount curBalance = pwallet->GetBalance();
+    CAmount curBalance = pwallet->GetBalance(ISMINE_SPENDABLE, 0, coin_control.m_allow_dirty_addresses);
```
I think this change should be in the _wallet/rpc: add allow_dirty option to sendtoaddress_ commit (not the _wallet/rpc: add include_dirty flag to getbalance command_ commit)

kallewoof replied:
Oops, yep. Thanks.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,113 @@ static UniValue loadwallet(const JSONRPCRequest& request)
 }
 
-static UniValue createwallet(const JSONRPCRequest& request)
+static std::map<std::string,WalletFlags> walletflagmap{
```
Can we define these in `wallet.h`, so there's only one place to change when adding wallet flags?

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,113 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+    {"disable_private_keys", WALLET_FLAG_DISABLE_PRIVATE_KEYS},
+};
+
+static UniValue getwalletflags(const JSONRPCRequest& request)
```
I'd prefer not to add this new method. Can we just add a `wallet_flags` field to the `getwalletinfo` return object?

kallewoof replied:
Removing it. There's already a "private_keys_enabled" and "avoid_reuse" entry in the returned value, so I don't think a wallet_flags is even necessary.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,113 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Unknown wallet flag %s", flag));
+    }
+
+    if (walletflagmap.at(flag) >= (1ULL << 32)) {
```
I don't think this is always necessarily going to be true. The meaning of the upper section/lower section flags are which we can safely ignore if we don't know the meaning of (ie if we open a newer wallet on an older client), not whether the flags are mutable or not.

kallewoof replied:
Ahh. Got it.

jnewbery commented on test/functional/test_runner.py:
```diff
@@ -94,6 +94,7 @@
     'rpc_getchaintips.py',
     'interface_rest.py',
     'mempool_spend_coinbase.py',
+    'feature_avoidreuse.py',
```
Should be `wallet_avoidreuse.py`

kallewoof replied:
Renamed

jnewbery reviewed (commented):
Concept ACK. Some additional high-level feedback inline.

kallewoof committed:
wallet/rpc: add 'avoid_reuse' option to RPC commands

createwallet, getbalance, getwalletinfo, listunspent, sendtoaddress

rpc/wallet: listunspent include reused flag and show reused utxos by default

kallewoof committed:
test: add test for avoidreuse feature

kallewoof commented:
@jnewbery Thanks for review! I believe I've addressed all your nits.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,57 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+
+    if (pwallet->IsWalletFlagSet(walletflagmap.at(flag))) return false;
+
+    pwallet->SetWalletFlag(walletflagmap.at(flag), value);
```
I don't think this is sufficient for enabling 'avoid_reuse'. Consider:

- wallet is created without the `avoid_reuse` flag
- wallet sends from address
- wallet receives new coins to used address. These are NOT marked as dirty because the `avoid_reuse` flag is not set
- `avoid_reuse` is enabled with the `setwalletflag` RPC
- `getbalance`, `sendtoaddress`, etc will all use the reused address coins because they're not set as dirty.

kallewoof replied:
Right ... the idea is that the user has to actually call 'rescan' for it to become retroactive. This was explained when this was a command line option, I think, but the explanation vanished with the change to remove the option. Not sure where would be a good place to explain this. Perhaps there should be a 'caveats' map for flags which are returned when setting flags?

jnewbery commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the avoid_reuse feature."""
```
It'd be nice if this tested the `setwalletflag` RPC too.

kallewoof replied:
It sort of does, actually. I assumed there were tests elsewhere, but apparently not. I will add complementary tests for setwalletflag.

jnewbery reviewed (commented):
Lots of comments.

I haven't thought too much about the logic for making coins dirty/checking dirtiness or the test coverage.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -339,9 +339,10 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
         throw std::runtime_error(
-            "sendtoaddress \"address\" amount ( \"comment\" \"comment_to\" subtractfeefromamount replaceable conf_target \"estimate_mode\")\n"
+            std::string() +
+            "sendtoaddress \"address\" amount ( \"comment\" \"comment_to\" subtractfeefromamount replaceable conf_target \"estimate_mode\" allow_dirty )\n"
```
The reason why we've stuck with dirty so far is because of the negative connotation. Note that a coin only becomes dirty if

* you received to an address
* you spent everything from that address
* you received to the same address afterwards

Imagine walking through mud on the way home. As you go inside, you leave a trail behind on the walkway up to and to the inside of your house, even if you prudently take the shoes off at the entrance.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,57 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+            "\nChange the state of the given wallet flag for a wallet.\n"
+            "\nArguments:\n"
+            "1. flag    (string, required) The name of the flag to change. Current available flags: " + flags + "\n"
+            "2. value   (boolean, optional, default=true) The new state.\n"
```
Oops, thanks! Fixed.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,57 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+            "1. flag    (string, required) The name of the flag to change. Current available flags: " + flags + "\n"
+            "2. value   (boolean, optional, default=true) The new state.\n"
+            "\nResult:\n"
+            "true|false (boolean) True or false, reflecting whether the flag was modified (true), or (false) whether it was already set to the given state\n"
```
I think being told that your action did nothing vs your action did something is a good UI, personally.

Perhaps it should actually *throw* instead of returning false, though...

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,57 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Unknown wallet flag %s", flag));
+    }
+
+    if (pwallet->IsWalletFlagSet(walletflagmap.at(flag))) return false;
```
D'oh, thanks. Fixed.

kallewoof committed:
wallet/rpc: add setwalletflag RPC and MUTABLE_WALLET_FLAGS

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2540,15 +2554,57 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+
+    if (pwallet->IsWalletFlagSet(walletflagmap.at(flag))) return false;
+
+    pwallet->SetWalletFlag(walletflagmap.at(flag), value);
```
Should it be possible to *disable* it?

kallewoof committed:
wallet: enable avoid_reuse feature

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,126 @@
+        assert_equal(self.nodes[1].getwalletinfo()["avoid_reuse"], True)
+
+        # Stop and restart node 1
+        self.stop_node(1)
```
I kind of want the node to go down completely so we don't get some weird case where a flag persists when reloading wallet but not when restarting node. Even if that sounds like an impossibility.

ken2812221 commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,146 @@
+        self.setup_clean_chain = False
+        self.num_nodes = 2
+
+    def reset_balance(self, node, discardaddr):
```
Should skip this if wallet disabled.
```python3
    def skip_test_if_missing_module(self):
        self.skip_if_no_wallet()
```

Sjors reviewed (commented):
Concept ACK, code is getting close.

Can you update the PR name and description to describe the latest behavior? My understanding:

* in order to avoid reuse for a fresh wallet, use `createwallet false true`
* for an existing wallet use `setwalletflag avoid_reuse true`
  * `setwalletflag` can't be used on  `disable_private_keys` because it's immutable
* `sendtoaddress` can override the flag using `allow_dirty`

Issues:
* the flag name `disable_private_keys` in `WALLET_FLAG_MAP` is inconsistent with the output of `getwalletinfo` (`private_keys_enabled`)
* the use of both `allow_dirty` and `avoid_reuse` is confusing. I find `avoid_reuse` more clear.
* why does `sendtoaddress` have an `allow_dirty` option, but e.g. `sendmany` doesn't?
* why does `getbalance` have `include_dirty`, but similar calls don't? (in particular `getunconfirmedbalance` and `getreceivedbylabel`)
* `setwalletflag avoid_reuse false` should mark used destinations as dirty or warn that it doesn't

Suggestions:
* don't tease the user by listing  `disable_private_keys` in the help, or add a separate "Flags which can't be changed: "
* as pointed out earlier, `listtransactions` and friends should include the dirty flag and allow excluding those, for UI purposes, but this can wait

practicalswift commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,149 @@
+        self.setup_clean_chain = False
+        self.num_nodes = 2
+
+    def reset_balance(self, node, discardaddr):
```
Nit: This method doesn't use `self` and could be a function instead?

kallewoof replied:
Good point. Turned into function.

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,149 @@
+
+        self.nodes[0].generate(110)
+        self.sync_all()
+        self.reset_balance(self.nodes[1], self.nodes[0].getnewaddress())
```
That's a good point, but `getreceivedbyaddress` has not been updated to handle the `avoid_reuse` flag yet. Perhaps it's worth doing that and switch, but it's also reassuring to ensure no weird balance appeared from some unexpected address.

Sjors replied:
Multiple wallets per node might be another route.

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,149 @@
+        # Unload temp wallet
+        self.nodes[1].unloadwallet(tempwallet)
+
+    def test_fund_send_fund_send(self):
```
The way I see it is, the two tests above it will catch weird issues that could cause the test to fail in random places, so those tests are run first. The intention is to order tests chronologically.

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,149 @@
+
+        self.nodes[1].sendtoaddress(retaddr, 4)
+
+    def test_fund_send_fund_senddirty(self):
```
OK!

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2439,15 +2466,65 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+
     if (request.fHelp || request.params.size() < 1 || request.params.size() > 2) {
+        std::string flags = "";
+        for (auto& it : WALLET_FLAG_MAP)
```
Could use `std::accumulate`?

luke-jr commented:
Dirty should be set when receiving, not when the received coins are then spent.

kallewoof commented:
@luke-jr 

> Dirty should be set when receiving, not when the received coins are then spent.

This is done in `AddToWallet`, so it looks like it is doing what you are asking for:

https://github.com/bitcoin/bitcoin/blob/dbfb209900cb6d4e6af3c0eb6ffa96f20f1dd123/src/wallet/wallet.cpp#L880-L886

The test `test_fund_send_fund_send` ensures that this behavior is the case, I think. Is it missing something?

MeshCollider commented on src/wallet/wallet.h:
```diff
@@ -125,11 +125,22 @@ enum WalletFlags : uint64_t {
 
+    // will categorize coins as clean (not reused) and dirty (reused), and handle
+    // them with privacy considerations in mind
+    WALLET_FLAG_AVOIDREUSE = (1ULL << 0),
```
nit: `WALLET_FLAG_AVOID_REUSE` instead of `WALLET_FLAG_AVOIDREUSE`

Sjors replied:
Agree with the nit, and also that it's a nit :-)

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -734,7 +735,11 @@ static UniValue getbalance(const JSONRPCRequest& request)
-    return ValueFromAmount(pwallet->GetBalance(filter, min_depth));
+    bool avoid_reuse = request.params[3].isNull()
+        ? pwallet->IsWalletFlagSet(WALLET_FLAG_AVOIDREUSE)
+        : request.params[3].get_bool();
```
Above help text says `(only applicable if the \"avoid reuse\" wallet flag is enabled)`, so why do you have this case? Shouldn't we just error immediately if it is set without `pwallet->IsWalletFlagSet(WALLET_FLAG_AVOIDREUSE)`? Same with the other RPCs

kallewoof replied:
Code now throws an error if user does `avoidreuse=true` for a wallet which does not have `WALLET_FLAG_AVOID_REUSE` enabled.

MeshCollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -555,13 +561,14 @@ static UniValue getreceivedbyaddress(const JSONRPCRequest& request)
             "\nArguments:\n"
             "1. \"address\"         (string, required) The bitcoin address for transactions.\n"
             "2. minconf             (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. avoid_reuse         (boolean, optional) Forbid inclusion of addresses which have been marked as 'dirty'.\n"
```
Help texts for all the RPC changes in this PR could do with some more consistency, e.g. in how the default is output (this one doesn't output the default at all, whereas `sendtoaddress` will decide whether to output `default=true` or `unavailable`)

meshcollider reviewed (commented):
Concept ACK / light code review https://github.com/bitcoin/bitcoin/pull/13756/commits/a2d136dc46593f6e15c038f2fb497596617ee73e

Haven't looked at the test yet. A couple of comments inline.

The split into multiple commits is nice but perhaps a little too much, some squashing would be good e.g. "wallet: add MUTABLE_WALLET_FLAGS" -> "wallet/rpc: add setwalletflag RPC command for modifying wallet flag post creation", etc.

kallewoof commented:
@MeshCollider Thanks much for the review!

> The split into multiple commits is nice but perhaps a little too much, some squashing would be good e.g. "wallet: add MUTABLE_WALLET_FLAGS" -> "wallet/rpc: add setwalletflag RPC command for modifying wallet flag post creation", etc.

I will try to squash stuff. Agree about the two commits you mentioned.

Sjors commented:
I wrote earlier:
> `setwalletflag avoid_reuse false` should mark used destinations as dirty or warn that it doesn't

I meant to say `setwalletflag avoid_reuse true` should go through existing transactions in the wallet and mark things dirty as needed. Or the help text and/or release notes should warn that it doesn't.

`getbalance "*" 0 false false` ignores unconfirmed transactions (pre-existing bug?)

`getreceivedbyaddress tb1q... 0 true` discards the first transaction on the dirty address. That seems non-intuitive.

A followup PR could add a "Avoid Reuse" checkbox to the coin selection dialog in the GUI to reveal dirty addresses. Another PR could add a Avoid Reuse checkbox to the Options->Wallet dialog, which would perform `setwalletflag avoid_reuse true`.

MeshCollider commented:
@Sjors getbalance is preexisting yep, see #14602

kallewoof committed:
doc: release notes for avoid_reuse

kallewoof commented:
@Sjors Thanks for testing and for valuable feedback.

> > setwalletflag avoid_reuse false should mark used destinations as dirty or warn that it doesn't
> 
> I meant to say setwalletflag avoid_reuse true should go through existing transactions in the wallet and mark things dirty as needed. Or the help text and/or release notes should warn that it doesn't.

I somehow forgot about this, apologies -- I've added "caveats" to the setwalletflags, and also added release notes indicating a rescan is required.

> getreceivedbyaddress tb1q... 0 true discards the first transaction on the dirty address. That seems non-intuitive.

Thinking about this, I actually don't really see why `getreceivedbyaddress` should even care about reuse. It should just show everything that was received. Dropping that commit.

Sjors reviewed (approved):
tACK 3bff2a5 on macOS 10.14.1

I would be good to get someone with deep wallet knowledge like @sipa to specifically check d4e108f6bbd99264a739455f9e46f7068acb01d0, which is where the used tracking happens.

kallewoof commented:
@Sjors https://github.com/bitcoin/bitcoin/pull/13756#discussion_r236677538 I can't see what nit this is referring to -- is it the `AVOID_REUSE` one? Then cool, as I believe I addressed that.

Sjors commented:
utACK 23a15c5

Changes since my last review: rebase, some commits reorganised, new RPC help syntax, cosmetic changes: `PREV=3bff2a5 N=14 && git range-diff `git merge-base --all HEAD $PREV`...$PREV HEAD~$N...HEAD`

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -4151,16 +4218,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
-        } else {
+        if ((wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
+            (wallet_creation_flags & WALLET_FLAG_BLANK_WALLET)) {
+            InitError(_("Wallet cannot have both DISABLE_PRIVATE_KEYS and BLANK_WALLET flags"));
```
Nit: this new check (if you keep it, see ff7fe3cf817f787ee8d929b73522d2e9c9da5681), as well as setting all flags in one go (below), deserves a separate commit. I do like the cleanup.

kallewoof replied:
I assumed setting both was an error, but realized tests rely on being able to set both. I think I screwed up the logic though (the original code did if/else, which would actually not set one of the flags if the other one was set). Will address now.

Sjors replied:
Actually setting both is fine, so this check needs to be dropped.
* a blank wallet with private keys enabled can import or generate new private keys
*  a blank wallet with private keys disabled can only import public keys

kallewoof replied:
Yes, setting both is fine, but the code pre-this-PR will actually ignore the blank wallet flag if the disable private keys flag is set:

https://github.com/bitcoin/bitcoin/blob/9c93f5d9fc93df2120998e8383bc972b738f3ff5/src/wallet/wallet.cpp#L4142-L4151

Sjors reviewed (approved):
~re-utACK 79bf9f6~, modulo squashing the new fixup commits:
* rebased on blank wallet #15226
* slight cleanup of `CreateWalletFromFile`
* modernised RPC help syntax

kallewoof commented:
Squashed. @Sjors I believe the code now acts the same way as master, i.e.

* blank_wallet → blank_wallet
* disable_privkeys → disable_privkeys
* blank_wallet | disable_privkeys → disable_privkeys

Sjors commented:
@kallewoof you're right, it's the same as on master. utACK ec26b2e

kallewoof commented:
@Sjors The big question is, does it even matter if both flags are set? I would love to not have those 5 lines of cruft in there, if they don't make a difference. I should hunt down whoever wrote that code and ask, I guess. (And/or verify myself.)

Sjors commented:
I'm not sure, but perhaps it's better dealt with in a followup to this as well as #14021. cc @achow101

It seems to me that when you create a wallet without private keys, as long as there's no additional argument to feed the new wallet with public keys, it's by definition blank. Problem is that the RPC needs a default and we probably don't want to make that too complicated. A sane default for `blank` is true when private keys are disabled, whereas otherwise the sane default for `blank` is false.

Sjors commented:
utACK rebased 74581ff

kallewoof commented:
Rebased, and refactored code based on #15780 cachable accounts (note: first commit here is from that PR).

There is now a isminetype called "ISMINE_USED" which is set for the 'allow reuse' caches.

Sjors reviewed (changes requested):
tACK 823205a _modulo params bug_ (inline). Compared to my previous tACK 74581ff:
* some commits were combined
* `ISMINE_USED` added (the cache logic in `GetAvailableCredit` could use extra eyes)

For a followup: this flag works best when combined with `-avoidpartialspends`, as suggested by @kallewoof [earlier](https://github.com/bitcoin/bitcoin/pull/12257#issuecomment-360051914). There's no harm as long as they are combined in a single transaction. If the user forgets to set `-avoidpartialspends` they'll miss the opportunity to spend duplicates in a safe way. We could interpret `-avoidpartialspends` (#12257) as `true` for wallets with `avoid_reuse` flag.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -938,6 +977,14 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
 
     uint256 hash = wtxIn.GetHash();
 
+    if (IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE)) {
```
We do not mark destinations as used unless the user turns out the functionality. This is the reason why people must rescan the chain if they turn `avoid_reuse` on.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2596,6 +2684,10 @@ static UniValue createwallet(const JSONRPCRequest& request)
         flags |= WALLET_FLAG_BLANK_WALLET;
     }
 
+    if (!request.params[2].isNull() && request.params[2].get_bool()) {
```
Wow. I need to write better tests, it looks like.

Sjors replied:
It was probably a rebase glitch. The tests often use named params.

kallewoof commented:
@Sjors Very good point about `-avoidpartialspends`. I can't believe I forgot about that. I think as you say this flag should be enabled by default if `avoid_reuse` is set in the wallet.

kallewoof committed:
wallet: enable avoid_partial_spends by default if avoid_reuse is set

Sjors commented:
Don't forget to update the bitcoind docs for this new param behavior.

kallewoof commented:
@Sjors Right, thanks. Done. I think a follow-up PR should convert `-avoidpartialspends` into a wallet flag that is auto-enabled when `avoid_reuse` flag is set/enabled.

kallewoof committed:
bitcoind: update -avoidpartialspends description to account for auto-enable for avoid_reuse wallets

Sjors commented:
utACK 69ec084, compared to my previous utACK:
* renamed  `SetUsedState` to `SetUsedDestinationState`, and `IsUsed` to `IsUsedDestination`
* param index fixed (@MarcoFalke it would be nice if we didn't need brittle stuff like `request.params[3]` in the RPC)
* improved documentation

jnewbery reviewed (commented):
Sorry for being so negligent in reviewing this. I've gone through all the commits again and have a few comments.

I think the `avoid_reuse` parameter should also be added to `sendmany`, but that could be done in a follow-up PR and shouldn't hold this up any more.

Sjors commented on src/wallet/init.cpp:
```diff
@@ -39,7 +39,7 @@ const WalletInitInterface& g_wallet_init_interface = WalletInit();
 {
     gArgs.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), false, OptionsCategory::WALLET);
-    gArgs.AddArg("-avoidpartialspends", strprintf("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it after spending from it), but may result in slightly higher fees as suboptimal coin selection may result due to the added limitation (default: %u)", DEFAULT_AVOIDPARTIALSPENDS) ...
+    gArgs.AddArg("-avoidpartialspends", strprintf("Group outputs by address, selecting all or none, instead of selecting on a per-output basis. Privacy is improved as an address is only used once (unless someone sends to it after spending from it), but may result in slightly higher fees as suboptimal coin selection may result due to the added limitation (default: %u%s)", DEFAULT_AVOIDPARTIALSPEND ...
```
Are you suggesting #ifdef? Because we might switch this default at some point.

jnewbery replied:
I just think it's weird to have runtime programming logic to construct a string where we already know what that string should be. If the concern is about this help text not being updated if `DEFAULT_AVOIDPARTIALSPENDS` changes, then I think it would be better to just leave a comment where that constant is defined saying "If this is changed to true, update the help text for -avoidpartialspends"

kallewoof replied:
I am going to just remove the conditional part of this and have it always say `"(always enabled for wallets with \"avoid_reuse\" enabled)"`. That should resolve the issue, I think?

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -1533,10 +1580,14 @@ bool CWallet::CanGetAddresses(bool internal)
-    m_wallet_flags |= flags;
+    if (!flag || (flag & (flag-1))) {
+        // flag has no bits set or more than one bit set
+        throw std::runtime_error(std::string(__func__) + ": flag format invalid (not a single bit set)");
```
Makes sense, changed.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -4057,16 +4138,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
-        } else if (wallet_creation_flags & WALLET_FLAG_BLANK_WALLET) {
-            walletInstance->SetWalletFlag(WALLET_FLAG_BLANK_WALLET);
-        } else {
+        if ((wallet_creation_flags & WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
```
@jnewbery Existing code would set at most one of the two flags (disable privkeys or blank_wallet), and would not set any other wallet flags. As I am introducing more wallet flags, I needed those to be set, but because the existing code will *ignore* the blank_wallet flag if the *disable privkeys* flag is set, I had to do some tweakery to not modify the existing behavior.

If everyone agrees that suddenly setting both disable_privkeys AND blank_wallet is harmless, this code can be simplified down to just doing `->SetWalletFlags(...)` without the if block, but I didn't feel confident making that decision myself.

jnewbery replied:
Ah. I see now. Yes, this makes sense, although it's a little confusing the way that the generic `setWalletFlags()` is called between two blocks of blank wallet specific code.

I've done some grepping and it seems to me that it's fine for `WALLET_FLAG_BLANK_WALLET` and `WALLET_FLAG_DISABLE_PRIVATE_KEYS` to be set on the same wallet. @achow101 should know for sure.

achow101 replied:
It's fine to set both `WALLET_FLAG_BLANK_WALLET` and `WALLET_FLAG_DISABLE_PRIVATE_KEYS` on a wallet, though redundant. `WALLET_FLAG_BLANK_WALLET` will be unset whenever something is imported.

ryanofsky replied:
I wish the `BLANK_WALLET` flag was called something more like `WALLET_IS_ALREADY_INITIALIZED` or `WALLET_DO_NOT_REINITIALIZE` since the only effect of the flag is to skip the `fFirstRun` code path when the wallet is loaded. I think the term "blank wallet" is good as a createwallet option name, but less good as a flag name.

kallewoof replied:
Removed the redundant if block.

kallewoof commented on src/wallet/wallet.cpp:
```diff
@@ -3043,6 +3116,14 @@ bool CWallet::CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::ve
             }
+
+            if (used) {
+                // resulting outputs are all used (if an output is not ours, it
```
This is just a convenient place to mark-up your own UTXO:s, e.g. change address and the like. This part can be removed and the `AddToWallet*` methods will handle it a little later. I'm actually going to remove it, as it's confused myself more than once in the past, and now, you as well.

kallewoof replied:
I investigated further, and this is indeed a bug. I updated the tests to ensure that the balance was updated correctly, and for the case where this `if` block was present, the non-dirty balance would become 0 when it should be 5.
```diff
diff --git a/test/functional/wallet_avoidreuse.py b/test/functional/wallet_avoidreuse.py
index b7b3eae34..2964d0868 100755
--- a/test/functional/wallet_avoidreuse.py
+++ b/test/functional/wallet_avoidreuse.py
@@ -105,11 +105,15 @@ class AvoidReuseTest(BitcoinTestFramework):

         self.nodes[1].sendtoaddress(address=retaddr, amount=10, avoid_reuse=False)

+        # node 1 should now have about 5 btc left (for both cases)
+        assert_approx(self.nodes[1].getbalance(), 5, 0.001)
+        assert_approx(self.nodes[1].getbalance(avoid_reuse=False), 5, 0.001)
+
     def test_fund_send_fund_send(self):
         '''
         Test the simple case where [1] generates a new address A, then
         [0] sends 10 BTC to A.
-        [1] spends 5 BTC from A. (leaving roughly 4 BTC useable)
+        [1] spends 5 BTC from A. (leaving roughly 5 BTC useable)
         [0] sends 10 BTC to A again.
         [1] tries to spend 10 BTC (fails; dirty).
         [1] tries to spend 4 BTC (succeeds; change address sufficient)
@@ -138,5 +142,9 @@ class AvoidReuseTest(BitcoinTestFramework):

         self.nodes[1].sendtoaddress(retaddr, 4)

+        # node 1 should now have about 1 btc left (no dirty) and 11 (including dirty)
+        assert_approx(self.nodes[1].getbalance(), 1, 0.001)
+        assert_approx(self.nodes[1].getbalance(avoid_reuse=False), 11, 0.001)
+
 if __name__ == '__main__':
     AvoidReuseTest().main()
```
For case with `if (used) [...]` present:
```
2019-05-20T03:18:33.609000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/me/path/bitcoin/test/functional/test_framework/test_framework.py", line 194, in main
    self.run_test()
  File "./wallet_avoidreuse.py", line 45, in run_test
    self.test_fund_send_fund_senddirty()
  File "./wallet_avoidreuse.py", line 109, in test_fund_send_fund_senddirty
    assert_approx(self.nodes[1].getbalance(), 5, 0.001)
  File "./wallet_avoidreuse.py", line 17, in assert_approx
    raise AssertionError("%s < [%s..%s]" % (str(v), str(vexp - vspan), str(vexp + vspan)))
AssertionError: 0E-8 < [4.999..5.001]
```

jnewbery replied:
oh, I think I understand why this code was here now. In the test you've just shown a way of turning coins from dirty to clean by creating a tx that uses up the dirty input and makes a clean change output.

I can see arguments for why you might want to consider it dirty, but on balance I think you've got the right logic now. The change output is a fresh address, so shouldn't be considered dirty.

kallewoof replied:
Yeah, I think I got the logic mixed up in there when I wrote that block. Thanks for pointing that out!

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2638,16 +2660,84 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+    pwallet->SetWalletFlag(flag, value);
+
+    if (flag && WALLET_FLAG_CAVEATS.count(flag)) {
+        res.pushKV("caveats", WALLET_FLAG_CAVEATS.at(flag));
```
I like `success` better than simply saying what the flag was changed to, because it better addresses the case where the user is targeting the wrong wallet. For example, I may ssh into a server with a node on it, and my intention is to turn a wallet flag on, but in this scenario, I ssh'd into the wrong machine and/or picked the wrong wallet (in the case of a node with multiple wallets).

With `success=<flag was actually changed>`, I would get a `success=false` here, which would hopefully lead me to discover my error. With `new_flag_value=true` here, I would get no such indication.

Perhaps both would be a good compromise here? Add `flag_name`, `flag_old_value`, `flag_new_value` to the results. Too verbose?

A caveat could apply to turning off a flag as well as turning it on, though, but as this is not the case right now, I'm going with "only show when turning flag to ON" approach.

As a non-native speaker, I don't think caveat is that weird a word, but will change it if people think it's hard to understand.

jnewbery replied:
> Add flag_name, flag_old_value, flag_new_value to the results. Too verbose?

Seems ok to me. Alternatively, you could just throw a `JSONRPCError` if the flag is already set to what was requested. That's the way we usually deal with input errors in the RPCs.

> As a non-native speaker, I don't think caveat is that weird a word

I'm going to claim that 'warning' is a [far more common word](https://books.google.com/ngrams/graph?content=caveat%2Cwarning&year_start=1800&year_end=2008&corpus=15&smoothing=7&share=&direct_url=t1%3B%2Ccaveat%3B%2Cc0%3B.t1%3B%2Cwarning%3B%2Cc0), and conveys the same meaning.

kallewoof replied:
Throwing seems sensible. Going with that and adding flag name and value (but skipping old value as it's now implied). Switching caveat to warning.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -386,6 +399,7 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1")
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"donation\" \"seans outpost\"")
             + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" true")
+            + HelpExampleCli("sendtoaddress", "\"1M72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 0.1 \"\" \"\" false true")
```
Yeah. Removed.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -722,7 +739,7 @@ static UniValue getbalance(const JSONRPCRequest& request)
     }
 
-    if (request.fHelp || (request.params.size() > 3 ))
+    if (request.fHelp || request.params.size() > 4)
```
I think a follow-up is a good idea here. I'll make an effort to write one soon after merge so this is fresh in people's heads.

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,150 @@
+        # Unload temp wallet
+        self.nodes[1].unloadwallet(tempwallet)
+
+    def test_fund_send_fund_senddirty(self):
```
I think I like the two separate, especially now that they ensure the balance updates are correct in each case. Look at the updated tests, and if you still think they should be merged, I can do so!

kallewoof commented:
@jnewbery Thanks for the review! It's 100% my responsibility to get people to review, so no worries at all. :)

jnewbery commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,150 @@
+    def test_fund_send_fund_senddirty(self):
+        '''
+        Test the same as test_fund_send_fund_send, except send the 10 BTC with
+        the avoid_reuse flag set. This means the 10 BTC send should succeed,
```
I think this needs to be updated to say "node 1 sends the 10BTC with the avoid_reuse flag set to false"

jnewbery commented:
Looking better. Thanks for the quick response to my review. I had just a few more comments inline.

kallewoof commented:
@jnewbery Thanks for follow-up review! I believe I addressed everything you brought up.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -3032,6 +3091,20 @@ bool CWallet::CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::ve
+            if (IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE)) {
+                for (const auto& vin : wtxNew.tx->vin) {
+                    if (IsUsedDestination(vin.prevout.hash, vin.prevout.n)) {
+                        used = true;
```
I don't understand what this code block is doing. You have a boolean `used` which is initially set to false and then may be updated to true, but you never use the value of `used` anywhere.

kallewoof replied:
That was leftovers from the incorrect code that I removed earlier. Thanks, removed!

jnewbery reviewed (commented):
Getting close!

I think we should update `listunspent` to take a `avoid_reuse` flag, or at the very least default to showing coins that are reusing addresses. I was testing this manually and got very confused when `listunspent` stopped showing coins in my wallet that I knew existed.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2639,6 +2660,72 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+                "\nChange the state of the given wallet flag for a wallet.\n",
+                {
+                    {"flag", RPCArg::Type::STR, RPCArg::Optional::NO, "The name of the flag to change. Current available flags: " + flags},
+                    {"state", RPCArg::Type::BOOL, /* default */ "true", "The new state."},
```
Changed to `value`.

kallewoof commented:
@jnewbery Thanks a lot for all the feedback. I updated `listunspent` to (1) show reused UTXOs and (2) show a flag "reused" if the wallet has "avoid_reuse" enabled.

I also updated the tests to check that the listunspent results match what you would expect after each send.

kallewoof committed:
wallet: rename g_known_wallet_flags constant to KNOWN_WALLET_FLAGS

jnewbery commented:
utACK 3ae67c42386c7c34b1e4d6bafe0d2233ae4523ea

@achow101's comment here https://github.com/bitcoin/bitcoin/pull/13756#discussion_r287415117 could be addressed in a follow-up PR.

jnewbery commented:
utACK 66f3e9780d916d9983bab1c153ab4f303d55bbd0

kallewoof commented:
Removed secondary parameter to `SetWalletFlag` and restored the `UnsetWalletFlag` method, and using that instead now.

jnewbery commented:
ACK 5ebc6b0eb

(Reran the rebase from 66f3e97 myself and checked the diff)

laanwj commented:
Concept and code-review ACK f3855781fdc52382b7ccfd0060623fa8ab1541b5

Would be nice if this had final sign-off by @meshcollider as wallet maintainer before merge.

Thanks for working on privacy!
```
//! Forbids inclusion of dirty (previously used) addresses
```
Hehe. If this use of "dirty addresses" would catch on, maybe people would start avoiding re-using addresses out of a sense of hygiene. One can dream, right.
But agree with @Sjors that "dirty payments" is better to avoid—it's kind of loaded.

kallewoof commented:
@laanwj https://github.com/bitcoin/bitcoin/commit/f3855781fdc52382b7ccfd0060623fa8ab1541b5 is not the right commit, FYI!

jtimon commented:
Since I don't know the wallet code all that well, it would take a lot of time to be able to properly review this, but concept ACK.

fanquake commented:
@achow101 Might also want to take a final look?

laanwj commented:
> @laanwj f385578 is not the right commit, FYI!

Thanks for noticing, fixed, that was a copy/paste error I was copying a lot of commit hashes for #16200.

meshcollider commented:
Code review ACK https://github.com/bitcoin/bitcoin/pull/13756/commits/5ebc6b0eb267e0552c66fffc5e5afe7df8becf80

Only lightly reviewed the test (8f2e208f7c0468f9ba92bc789a698281b1c81284) but the actual code looks good, thank you @kallewoof for being so proactive in addressing review comments and rebasing! About time this finally gets merged. I'll wait a day for Andrew to review but otherwise I'll merge this tomorrow.

meshcollider commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -0,0 +1,217 @@
+        assert_raises_rpc_error(-8, "Wallet flag is already set to false", self.nodes[0].setwalletflag, 'avoid_reuse', False)
+        assert_raises_rpc_error(-8, "Wallet flag is already set to true", self.nodes[1].setwalletflag, 'avoid_reuse', True)
+
+    def test_immutable(self):
```
In addition to the above TODO, I also suggest this be moved to a different file later (its independent of the avoid-reuse tests)

kallewoof replied:
Makes sense to me.

achow101 commented on src/script/ismine.h:
```diff
@@ -20,7 +20,9 @@ enum isminetype
     ISMINE_SPENDABLE  = 1 << 1,
+    ISMINE_USED       = 1 << 2,
     ISMINE_ALL        = ISMINE_WATCH_ONLY | ISMINE_SPENDABLE,
+    ISMINE_ALL_USED   = ISMINE_ALL | ISMINE_USED,
```
Is it really necessary to add new ismine types? AFAICT, these are only used as ismine filters which I think could just as easily be done with a boolean. `IsMine()` isn't returning these values, and `ISMINE_ALL_USED` isn't used anywhere.

kallewoof replied:
You get most of the combinations with the new "used" flag, because a bunch of calls have a with and without "avoid reuse" state... so I think this is the cleanest approach.

`ISMINE_ALL_USED` is indeed not used; I added it to indicate that `ISMINE_ALL` is in fact not include the used ones. It's sort of a comment, with the added side effect that smart IDEs will show it when giving suggestions.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2639,6 +2660,76 @@ static UniValue loadwallet(const JSONRPCRequest& request)
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2) {
```
nit: Use `RPCHelpMan help{...}`. See https://github.com/bitcoin/bitcoin/blob/master/src/rpc/rawtransaction.cpp#L74 for an example.

kallewoof replied:
Will address if rebase becomes necessary, otherwise will try to preserve ACKs and do a quick follow-up post-merge.

kallewoof replied:
I can do this but it would mean some unnecessary calculations when users are not asking for help:

https://github.com/bitcoin/bitcoin/blob/44d81723236114f9370f386f3b3310477a6dde43/src/wallet/rpcwallet.cpp#L2671-L2674

Still worth it?

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2830,6 +2928,9 @@ static UniValue listunspent(const JSONRPCRequest& request)
             "    \"witnessScript\" : \"script\" (string) witnessScript if the scriptPubKey is P2WSH or P2SH-P2WSH\n"
             "    \"spendable\" : xxx,        (bool) Whether we have the private keys to spend this output\n"
             "    \"solvable\" : xxx,         (bool) Whether we know how to spend this output, ignoring the lack of keys\n"
+            + (avoid_reuse ?
```
I don't think we should have help output that switches on the active wallet for that command. This breaks `help listunspent`.

kallewoof replied:
Why does it break?

achow101 replied:
You can do `help listunspent` without specifying `-rpcwallet` in order to just get the help for `listunspent` (and other commands). By making the help dependent on `-rpcwallet`, this makes the output of `help listunspent` less helpful because then the help sometimes will include `reused` and other times not. The help text should be consistent regardless of the other options specified. In general, all options and output should be displayed in the help even if some of those results will be omitted in actual command output. For example, `getaddressinfo` has a ton of extra fields that may or may not be part of the actual result, but all of them are listed in the help output.

kallewoof replied:
Ahh, I see what you're saying. Will rework into static output post-merge and/or if rebase becomes necessary.

achow101 commented:
ACK 5ebc6b0eb267e0552c66fffc5e5afe7df8becf80 modulo above nits

Reviewed the code and did a couple of manual tests.

Still not a fan of adding new isminetypes but I won't block this on that.

meshcollider commented:
The few remaining nits are not worth blocking merge on this, it has waited long enough. A follow-up PR to address them would be good.

Sjors commented:
I took a very brief `range-diff` look at the changes since my last review 69ec084ace48378c2eecd1b6568639ecfb50469b. Not too many changes. Congrats on the merge!

MarcoFalke commented on doc/release-notes-13756.md:
```diff
@@ -0,0 +1,39 @@
+- getbalance
+- sendtoaddress
+
+In addition, `sendtoaddress` has been changed to enable `-avoidpartialspends` when
```
Can you explain what that means? `sendtoaddress` is an RPC call, how would it enable a command line flag (`-avoidpartialspends`) of a running program?

kallewoof replied:
Changing wording to
```
In addition, `sendtoaddress` has been changed to always use `-avoidpartialspends`
when `avoid_reuse` is enabled, as it would otherwise risk using up the "wrong" UTXO
for an address reuse case.
```
Does that look better?

MarcoFalke commented on doc/release-notes-13756.md:
```diff
@@ -0,0 +1,39 @@
+These include:
+
+- createwallet
+- getbalance
```
What about `getbalances`?

kallewoof replied:
In #16239.

MarcoFalke commented on doc/release-notes-13756.md:
```diff
@@ -0,0 +1,39 @@
+a wallet will distinguish between used and unused addresses, and default to not
+use the former in coin selection.
+
+(Note: rescanning the blockchain is required, to correctly mark previously
```
Could remove `(Note: `, since all sentences in this doc are release *notes*

kallewoof replied:
Removed. I wanted to use it as in "Do note that [word of caution]", but it's no big deal.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -378,6 +389,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
             "       \"ECONOMICAL\"\n"
             "       \"CONSERVATIVE\""},
+                    {"avoid_reuse", RPCArg::Type::BOOL, /* default */ pwallet->IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE) ? "true" : "unavailable", "Avoid spending from dirty addresses; addresses are considered\n"
+            "                             dirty if they have previously been used in a transaction."},
```
This makes the help text change at run time. I'd prefer to make it static and explain it with something like "true if wallet falg is set, otherwise unavailable"

kallewoof replied:
Yep, fixed in #16239.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -733,6 +749,7 @@ static UniValue getbalance(const JSONRPCRequest& request)
                     {"dummy", RPCArg::Type::STR, RPCArg::Optional::OMITTED_NAMED_ARG, "Remains for backward compatibility. Must be excluded or set to \"*\"."},
                     {"minconf", RPCArg::Type::NUM, /* default */ "0", "Only include transactions confirmed at least this many times."},
                     {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false", "Also include balance in watch-only addresses (see 'importaddress')"},
+                    {"avoid_reuse", RPCArg::Type::BOOL, /* default */ pwallet->IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE) ? "true" : "unavailable", "Do not include balance in dirty outputs; addresses are considered dirty if they have previously been used in a transaction."},
```
Same here

kallewoof replied:
Good catch. Added to #16239.

maflcko reviewed (commented):
Looked at the documentation and left some comments
